WriteHDLUtils.py

```diff
@@ -828,15 +834,35 @@ def writeModuleInst_generic(module, hls_src_dir, f_writeTemplatePars,
                     foundMatch = f_matchArgPortNames(argname, portname)
 
                 if foundMatch:
+                    # Create temporary argument name as argname can be an array and have several matches
+                    tmp_argname = argname
+                    argname_is_array = (tmp_argname.find('[') != -1) # Check if array
+
+                    # Special case if argname is an array
+                    if argname_is_array:
+                        tmp_argname = tmp_argname.split('[')[0] # Remove "[...]"
+                        # Keep track of the array names and the number of array elements
+                        if tmp_argname in array_dict:
+                            array_dict[tmp_argname] += 1
+                        else:
+                            array_dict[tmp_argname] = 0
+                        # Add array index to the name as HLS implements one port for each array element
+                        # Temporary bodge to account for encoded index in fullmatch memories
+                        if tmp_argname == 'fullmatch':
+                            tmp_argname += "_" + str(decodeSeedIndex_MC(memory.inst))
+                        else:
+                            tmp_argname += "_" + str(array_dict[tmp_argname])
+
                     # Add the memory instance to the port string
+                    # Assumes a sorted memModuleList due to arrays?
                     if portname.find("in") != -1:
-                        string_mem_ports += writeProcMemoryRHSPorts(argname,memory)
+                        string_mem_ports += writeProcMemoryRHSPorts(tmp_argname,memory)
                     if portname.find("out") != -1:
-                        string_mem_ports += writeProcMemoryLHSPorts(argname,memory)
+                        string_mem_ports += writeProcMemoryLHSPorts(tmp_argname,memory)
+
                     # Remove the already added module and name from the lists
-                    memModuleList.remove(memory)
                     portNameList.remove(portname)
-                    break
```

Comment:
Rather than remove the break, could we conditionally break here if argname_is_array is false? Then we don't loop through the remaining memories looking for a port name that we have already found. Might save some run time, not sure how significant.

Reply:
Good idea! I'll add that.